File: src/test/js/formulas/statistical/MINTest.ts
import { assert } from "chai";
import { it, describe } from "../../testutils/TestUtils";
import { stub } from "sinon";
import { F7ExceptionName } from "../../../../main/js/errors/F7ExceptionName";
import { NAException } from "../../../../main/js/errors/NAException";
import { MIN } from "../../../../main/js/formulas/statistical/MIN";
import { Grid } from "../../../../main/js/models/common/Grid";
import { CollateralLookupFunction, LookupFunction } from "../../../../main/js/models/common/Types";
import { CommonModels } from "../../testutils/CommonModels";
import { assertF7ExceptionByName } from "../../testutils/TestUtils";

describe("MIN", function () {
  const GRID = Grid.builder()
    .add(0, 0, 22.1)
    .add(0, 1, 324.3)
    .add(0, 2, 22.2312223131232)
    .add(0, 3, 442309.4)
    .add(0, 4, 131289731)
    .build();
  const GRID_WITH_TEXT_VALUE = Grid.builder()
    .add(0, 0, 22.1)
    .add(0, 1, "Ignore me I guess.")
    .add(0, 2, 10)
    .build();
  const GRID_WITH_BLANK_VALUES = Grid.builder()
    .add(0, 0, 2)
    .add(0, 1, 3)
    .add(0, 2, 2)
    // Leaving gap in range to simulate null/blank values.
    .add(0, 100, 4)
    .build();

  it("should do min minimum value for numbers", function () {
    assert.deepEqual(MIN.SELF.run(null, 1, 2, 3, 4, 5, 6, 7, 8), 1);
    assert.deepEqual(MIN.SELF.run(null, 1, 2, 3, 4, 5, 0, 7, 8), 0);
    assert.deepEqual(MIN.SELF.run(null, 0), 0);
    assert.deepEqual(MIN.SELF.run(null, 2984723.99382), 2984723.99382);
    assert.deepEqual(MIN.SELF.run(null, 1, 2, -1000), -1000);
  });

  it("should return VALUE error when encountering text values outside ranges", function () {
    assertF7ExceptionByName(MIN.SELF.run(null, 1, 2, "Bad"), F7ExceptionName.VALUE);
  });

  it("should ignore text values inside of grids", function () {
    assert.deepEqual(MIN.SELF.run(null, GRID_WITH_TEXT_VALUE), 10);
  });

  it("should return 0 when the resulting filter range is all null", function () {
    assert.deepEqual(MIN.SELF.run(null, Grid.builder().build()), 0);
  });

  it("should consider booleans", function () {
    assert.deepEqual(MIN.SELF.run(null, true, false, 10), 0);
  });

  it("should ignore blank cells (null values)", function () {
    assert.deepEqual(MIN.SELF.run(null, GRID_WITH_BLANK_VALUES), 2);
  });

  it("should use lookup function for range queries", function () {
    const lookup = stub();
    const collateralLookup = stub();
    const F = new MIN(lookup as LookupFunction, collateralLookup as CollateralLookupFunction);
    lookup.withArgs(CommonModels.M22_RANGE).returns(GRID);
    assert.deepEqual(F.run(CommonModels.A1, CommonModels.M22_RANGE), 22.1);
    assert.isTrue(lookup.calledWith(CommonModels.M22_RANGE));
    assert.equal(lookup.callCount, 1);
    assert.isTrue(collateralLookup.notCalled);
  });

  it("should return error when argument lengths are wrong", function () {
    assert.deepEqual((MIN.SELF.run(null) as NAException).name, F7ExceptionName.NA);
  });
});

File: src/test/java/io/protobase/f7/formulas/statistical/MINTest.java
package io.protobase.f7.formulas.statistical;

import io.protobase.f7.errors.NAException;
import io.protobase.f7.errors.ValueException;
import io.protobase.f7.models.Grid;
import io.protobase.f7.testutils.TestFormula;
import org.junit.Test;

import static com.google.common.truth.Truth.assertThat;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoMoreInteractions;
import static org.mockito.Mockito.when;

public class MINTest extends TestFormula {
  @Test
  public void testApply_normal() {
    assertThat(MIN.SELF.apply(null, 1.0, 2.0, 3.0)).isEqualTo(1.0);
    assertThat(MIN.SELF.apply(null, 3.0, 1.0, 2.0)).isEqualTo(1.0);
    assertThat(MIN.SELF.apply(null, 3.0, 2.0, 1.0)).isEqualTo(1.0);
    assertThat(MIN.SELF.apply(null, 0.0)).isEqualTo(0.0);
    assertThat(MIN.SELF.apply(null, 0.0, -1000.0, -0.0111)).isEqualTo(-1000.0);
    assertThat(MIN.SELF.apply(null, 261822.0, 991741987.1, 1991.0e2)).isEqualTo(1991.0e2);
  }

  @Test
  public void testApply_numberLike() {
    assertThat(MIN.SELF.apply(null, 1.0, 2.0, "3.0")).isEqualTo(1.0);
    assertThat(MIN.SELF.apply(null, -1.0, -2.0, true)).isEqualTo(-2.0);
    assertThat(MIN.SELF.apply(null, -1.0, -2.0, false)).isEqualTo(-2.0);
    assertThat(MIN.SELF.apply(null, 1.0, 2.0, false)).isEqualTo(0.0);
    assertThat(MIN.SELF.apply(null, 2.0, true)).isEqualTo(1.0);
  }

  @Test
  public void testApply_grid() {
    assertThat(MIN.SELF.apply(null, Grid.builder().add(0, 0, 1.0).add(0, 1, "Bad").build(), 44.0)).isEqualTo(1.0);
  }

  @Test
  public void test_lookup() {
    Grid grid = Grid.builder()
        .add(0, 0, 1.0)
        .add(0, 1, 2.0)
        .add(0, 2, 3.0)
        .add(0, 3, 4.0)
        .add(0, 4, 5.0)
        .build();
    MIN F = new MIN(lookup, collateralLookup);
    when(lookup.apply(D1_TO_D5_RANGE)).thenReturn(grid);
    assertThat(F.apply(A1, D1_TO_D5_RANGE)).isEqualTo(1.0);
    verify(lookup).apply(D1_TO_D5_RANGE);
    verifyNoMoreInteractions(lookup);
  }

  @Test
  public void testApply_nonCoercableValuesInGridShouldBeIgnoreButNotInArgs() {
    assertThat(MIN.SELF.apply(null, Grid.builder().add(0, 0, 1.0).add(0, 1, "Bad").build(), "Bad")).isEqualTo(new ValueException());
  }

  @Test
  public void testApply_errorFromArgumentsMismatch() {
    assertThat(MIN.SELF.apply(null)).isEqualTo(new NAException());
  }
}